esp-hal/src/dma/mod.rs

```diff
@@ -2148,11 +2251,36 @@ pub unsafe trait DmaTxBuffer {
 /// The implementing type must keep all its descriptors and the buffers they
 /// point to valid while the buffer is being transferred.
 pub unsafe trait DmaRxBuffer {
+    /// A type providing operations that are safe to perform on the buffer
+    /// whilst the DMA is actively using it.
+    type View: DmaBufferView<DmaBuffer = Self>;
     /// Prepares the buffer for an imminent transfer and returns
     /// information required to use this buffer.
     ///
     /// Note: This operation is idempotent.
-    fn prepare(&mut self) -> Preparation;
+    fn prepare_mut(&mut self) -> Preparation;
+
+    /// This method allows users to let the buffer implementation know that
+    /// [Self::prepare] will likely be called next.
+    ///
+    /// If [Self::prepare] is expensive, implementations are strongly encouraged
+    /// to do the expensive part in this function where possible.
+    ///
+    /// Users are encouraged to call this method if there are spare CPU cycles
+    /// available. This is typically whilst waiting for a previous transfer
+    /// to complete.
+    ///
+    /// This method may never be called so do not rely on it for correct
+    /// function.
+    ///
+    /// Note: This operation is idempotent.
+    fn prepare_hint(&mut self) {
+        // This method is optional.
+    }
+
+    /// Prepares the buffer for an imminent transfer and returns
+    /// information required to use this buffer.
+    fn prepare(self) -> (Preparation, Self::View);
```

Comment:
Keep the old `prepare()` function that borrows and gives you a Preparation, and add a new `into_view()`. That way you don't need to modify the DMA API.

Reply:
That makes it more difficult to describe the safety rules. It's easier to say (and understand) "the buffers and descriptors must be valid between prepare and release" rather than say "they must be valid whilst the buffer is in use".

Reply:
And yet, it allows for oh so much more flexibility. I'll take that for the low-low price of 2 sentences in a documentation.